content/blog/2022/2022-06-30-configuring-cors-with-spring.md

```diff
@@ -0,0 +1,430 @@
+
+### Configure CORS using @CrossOrigin
+
+Here we will look at various ways in which our client application will be able to access the server resource without throwing CORS errors.
+
+#### Defining @CrossOrigin at the class level
+
+````java
+    @CrossOrigin(maxAge = 3600)
+    @RestController
+    @RequestMapping("cors-library/managed/books")
+    public class LibraryController {}
+````
+
+Here since we have defined @CrossOrigin
+ - All cross-origin requests will be accepted.
+ - Since maxAge = 3600, all pre-flight responses will be cached for 30 mins.
+
+#### Defining @CrossOrigin at method level
+
+````java
+    @CrossOrigin(origins = "http://localhost:4200", allowedHeaders = "Requestor-Type", exposedHeaders = "X-Get-Header")
+    @GetMapping
+    public ResponseEntity<List<BookDto>> getBooks(@RequestParam String type) {
+        HttpHeaders headers = new HttpHeaders();
+        headers.set("X-Get-Header", "ExampleHeader");
+        return ResponseEntity.ok().headers(headers).body(libraryService.getAllBooks(type));
+        }
+````
+
+ - Request coming from origin http://localhost:4200 only will be processed.
+ - If we expect only certain headers to be accepted, we can specify those headers in the allowedHeaders attribute. If the "Requestor-Type" header is not sent by the browser, the request will not be processed.
+ - If we set certain response headers, for the client application to be able to use them, we need to explicitly set the list of response headers to be exposed using the **exposedHeaders** attribute.
+
+#### Combination of @CrossOrigin at class and method levels
+
+````java
+    @CrossOrigin(maxAge = 3600)
+    @RestController
+    @RequestMapping("cors-library/managed/books")
+    public class LibraryController {
+
+        private static final Logger log = LoggerFactory.getLogger(LibraryController.class);
+
+        private final LibraryService libraryService;
+
+        public LibraryController(LibraryService libraryService) {
+            this.libraryService = libraryService;
+        }
+
+        @CrossOrigin(origins = "http://localhost:4200", allowedHeaders = "Requestor-Type")
+        @GetMapping
+        public ResponseEntity<List<BookDto>> getBooks(@RequestParam String type) {
+            HttpHeaders headers = new HttpHeaders();
+            headers.set("X-Get-Header", "ExampleHeader");
+            return ResponseEntity.ok().headers(headers).body(libraryService.getAllBooks(type));
+        }
+    }
+````
+By defining the annotation at both class and method levels
+    - its combined attributes will be applied to the methods i.e (origins, allowedHeaders, maxAge)
+
+### Enable CORS Configuration globally
+
+Instead of adding CORS to each of the resource separately, we could define a common CORS configuration that would apply to
+all resources defined. Here, we will use **WebMvcConfigurer** which is a part of the Spring Web MVC library.
```

Comment:
It's usually better to show the code first and _then_ explain it. Presenting the code first will give the reader the chance to understand it themselves and then to confirm it with the explanation afterwards. Also, it's just the more natural sequence to first see the code and then have it explained.

In this case it's not too bad, because the code is very short, so no need to change it here.

Reply:
Noted